web3/middleware/filter.py

```diff
@@ -0,0 +1,324 @@
+    1
+    >>> next(new_blocks)  # Latest block = 10
+    10
+    >>> next(new_blocks)
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+      StopIteration
+    >>>
+    """
+    _last = None
+    while True:
+        latest_block = web3.eth.blockNumber
+        if (latest_block is not None and
+                to_block is not None and
+                latest_block > to_block):
+            return
+        #  No new blocks since last iteration.
+        if _last is not None and _last == latest_block:
+            yield None
+        else:
+            yield latest_block
+        _last = latest_block
+
+
+def count(start=0, step=1):
+    """A counter you can update with send()
+    """
+    i = start
+    while True:
+        val = (yield i)
+        if val is not None:
+            i = val
+        else:
+            i += step
+
+
+def iter_latest_block_ranges(web3, from_block, to_block=None):
+    """Returns an iterator unloading ranges of available blocks
+
+    starting from `fromBlock` to the latest mined block,
+    until reaching toBlock. e.g.:
+
+
+    >>> blocks_to_filter = unload_latest_blocks(w3, 0, 50)
+    >>> next(blocks_to_filter)  # latest block number = 11
+    (0, 11)
+    >>> next(blocks_to_filter)  # latest block number = 45
+    (12, 45)
+    >>> next(blocks_to_filter)  # latest block number = 50
+    (46, 50)
+    """
+    _from_block = from_block
+    for _latest_block in iter_latest_block(web3, to_block):
+        if _latest_block is None:
+            yield (None, None)
+        else:
+            yield (_from_block, _latest_block)
+            _from_block = _latest_block + 1
+
+
+def get_logs_multipart(
+        web3,
+        startBlock,
+        stopBlock,
+        address,
+        topics,
+        max_blocks):
+    """Use to make large requests to ``eth_getLogs``
+
+    The getLog request is broken into multiple calls of the max number of blocks
+    ``max_blocks``.
+    """
+    chunks = block_range_chunker(startBlock, stopBlock, max_blocks)
+    for from_block, to_block in chunks:
+        params = {
+            'fromBlock': from_block,
+            'toBlock': to_block,
+            'address': address,
+            'topics': topics
+        }
+        yield web3.eth.getLogs(params)
+
+
+class RequestLogs:
+    def __init__(
+            self,
+            web3,
+            from_block=None,
+            to_block=None,
+            address=None,
+            topics=None):
+
+        self.address = address
+        self.topics = topics
+        self.web3 = web3
+        if from_block is None or from_block == 'latest':
+            self._from_block = web3.eth.blockNumber
+        else:
+            self._from_block = from_block
+        self._to_block = to_block
+        self.filter_changes = self._get_filter_changes()
+
+    @property
+    def from_block(self):
+        if self._from_block > self.web3.eth.blockNumber:
+            from_block = self.web3.eth.blockNumber
+
```

Comment:
Any reason for the extra whitespace?  Visually confusing to me but mostly because I'm not used to it.